pkg/apiserver/compression_io.go

```diff
@@ -0,0 +1,61 @@
+package apiserver
+
+import (
+	"compress/gzip"
+	"io"
+
+	"github.com/gin-gonic/gin"
+)
+
+type countingBodyWriter struct {
+	gin.ResponseWriter
+	writtenUncompressedBytes int
+}
+
+func (r *countingBodyWriter) WriteString(s string) (n int, err error) {
+	writtenBytes, err := r.ResponseWriter.WriteString(s)
+	r.writtenUncompressedBytes += writtenBytes
+
+	return writtenBytes, err
+}
+
+func (r *countingBodyWriter) Write(p []byte) (n int, err error) {
+	writtenBytes, err := r.ResponseWriter.Write(p)
+	r.writtenUncompressedBytes += writtenBytes
+
+	return writtenBytes, err
+}
+
+type countingBodyReader struct {
+	io.ReadCloser
+	readCompressedBytes int
+}
+
+func (r *countingBodyReader) Read(p []byte) (int, error) {
+	readBytes, err := r.ReadCloser.Read(p)
+	r.readCompressedBytes += readBytes
+
+	return readBytes, err
+}
+
+type gzipBodyReader struct {
+	body                  *countingBodyReader
+	reader                *gzip.Reader
+	readUncompressedBytes int
+}
+
+func (r *gzipBodyReader) Read(p []byte) (int, error) {
+	readBytes, err := r.reader.Read(p)
+	r.readUncompressedBytes += readBytes
+
+	return readBytes, err
+}
+
+func (r *gzipBodyReader) Close() error {
+	err := r.reader.Close()
+	if err != nil {
+		return err
+	}
+
+	return r.body.Close()
```

Comment:
this could be leaking if reader can't be closed, but i guess that is very unlikely?

Reply:
better safe than sorry, fixed